File: src/main/java/io/github/kingg22/api_vacunas_panama/persistence/repository/DosisRepository.java
package io.github.kingg22.api_vacunas_panama.persistence.repository;

import io.github.kingg22.api_vacunas_panama.persistence.entity.Dosis;
import io.github.kingg22.api_vacunas_panama.persistence.entity.Paciente;
import io.github.kingg22.api_vacunas_panama.persistence.entity.Vacuna;
import org.springframework.data.jpa.repository.JpaRepository;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public interface DosisRepository extends JpaRepository<Dosis, UUID> {

    Optional<Dosis> findTopByPacienteAndVacunaOrderByCreatedAtDesc(Paciente paciente, Vacuna vacuna);

    List<Dosis> findAllByPaciente_IdAndVacuna_IdOrderByCreatedAtDesc(UUID paciente, UUID id);

}

File: src/main/kotlin/io/github/kingg22/api/vacunas/panama/modules/vacuna/repository/DosisRepository.kt
package io.github.kingg22.api.vacunas.panama.modules.vacuna.repository

import io.github.kingg22.api.vacunas.panama.modules.paciente.entity.Paciente
import io.github.kingg22.api.vacunas.panama.modules.vacuna.entity.Dosis
import io.github.kingg22.api.vacunas.panama.modules.vacuna.entity.Vacuna
import io.github.kingg22.api.vacunas.panama.util.withSession
import io.quarkus.hibernate.reactive.panache.kotlin.PanacheRepositoryBase
import io.smallrye.mutiny.coroutines.awaitSuspending
import jakarta.enterprise.context.ApplicationScoped
import java.util.UUID

@ApplicationScoped
class DosisRepository : PanacheRepositoryBase<Dosis, UUID> {
    suspend fun findTopByPacienteAndVacunaOrderByCreatedAtDesc(paciente: Paciente, vacuna: Vacuna): Dosis? =
        withSession {
            find(
                "paciente = :paciente AND vacuna = :vacuna ORDER BY createdAt DESC",
                mapOf("paciente" to paciente, "vacuna" to vacuna),
            ).firstResult().awaitSuspending()
        }

    suspend fun findAllByPacienteIdAndVacunaIdOrderByCreatedAtDesc(paciente: UUID, id: UUID): List<Dosis> =
        withSession {
            list(
                "paciente.id = :paciente AND vacuna.id = :id ORDER BY createdAt DESC",
                mapOf("paciente" to paciente, "id" to id),
            ).awaitSuspending()
        }
}
